src/metabase/sync/sync_metadata/fks.clj

```diff
@@ -1,65 +1,101 @@
-                              :visibility_type    [:not= "retired"])]
-        {:source-field source-field
-         :dest-table   dest-table
-         :dest-field   dest-field}))))
+   [toucan2.core :as t2]
+   [toucan2.realize :as t2.realize]))
 
+(defn ^:private mark-fk-sql
+  "Returns [sql & params] for [[mark-fk!]] according to the application DB's dialect."
+  [{:keys [db-id
+           fk-table-name
+           fk-table-schema
+           fk-column-name
+           pk-table-name
+           pk-table-schema
+           pk-column-name]}]
+  (let [field-id-query (fn [db-id table-schema table-name column-name]
+                         {:select [[[:min :f.id] :id]]
+                          ;; Cal 2024-03-04: We use `min` to limit this subquery to one result (limit 1 isn't allowed
+                          ;; in subqueries in MySQL) because it's possible for schema, table, or column names to be
+                          ;; non-unique when lower-cased for some DBs. We have been doing case-insensitive matching
+                          ;; since #39679 so this preserves behaviour to avoid possible regressions.
+                          :from   [[:metabase_field :f]]
+                          :join   [[:metabase_table :t] [:= :f.table_id :t.id]]
+                          :where  [:and
+                                   [:= :t.db_id db-id]
+                                   [:= [:lower :f.name] (u/lower-case-en column-name)]
+                                   [:= [:lower :t.name] (u/lower-case-en table-name)]
+                                   [:= [:lower :t.schema] (some-> table-schema u/lower-case-en)]
+                                   [:= :f.active true]
+                                   [:not= :f.visibility_type "retired"]
+                                   [:= :t.active true]
+                                   [:= :t.visibility_type nil]]})
+        fk-field-id-query (field-id-query db-id fk-table-schema fk-table-name fk-column-name)
+        pk-field-id-query (field-id-query db-id pk-table-schema pk-table-name pk-column-name)
+        q (case (mdb.connection/db-type)
+            :mysql
+            {:update [:metabase_field :f]
+             :join   [[fk-field-id-query :fk] [:= :fk.id :f.id]
+                      ;; Only update if either:
+                      ;; - fk_target_field_id is NULL and the new target is not NULL
+                      ;; - fk_target_field_id is not NULL but the new target is different and not NULL
+                      [pk-field-id-query :pk]
+                      [:and
+                       [:or
+                        [:= :f.fk_target_field_id nil]
+                        [:not= :f.fk_target_field_id :pk.id]]]]
+             :set    {:fk_target_field_id :pk.id
+                      :semantic_type      "type/FK"}}
+            :postgres
+            {:update [:metabase_field :f]
+             :from   [[fk-field-id-query :fk]]
+             :join   [[pk-field-id-query :pk] true]
```

Comment:
Does Postgres not support the constraint going here for the statement, or is this a quirk for performance?